File: packages/core/lib/src/agent/mode.ts
/**
 * Mode management — controls the active mode (agent/admin) for a server
 * and admin certificate file operations.
 *
 * Ported from the Tauri desktop app's mode.rs.
 */

import { existsSync } from 'node:fs';
import { serverAdminP12Path, LAMASTE_DIR } from './platform.js';
import {
  loadServersRegistry,
  saveServersRegistry,
} from './server-registry.js';

// Re-export for convenience
export { serverAdminP12Path } from './platform.js';

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

export type ServerMode = 'agent' | 'admin';

// ---------------------------------------------------------------------------
// Mode management
// ---------------------------------------------------------------------------

/**
 * Set the active mode for a server ("agent" or "admin").
 */
export async function setServerMode(serverId: string, mode: ServerMode): Promise<void> {
  if (mode !== 'agent' && mode !== 'admin') {
    throw new Error("Mode must be 'agent' or 'admin'");
  }

  const servers = await loadServersRegistry();
  const server = servers.find((s) => s.id === serverId);
  if (!server) {
    throw new Error(`Server ${serverId} not found`);
  }

  server.activeMode = mode;
  await saveServersRegistry(servers);
}

/**
 * Get the active server's mode. Defaults to "agent" if no servers or no active server.
 */
export async function getServerMode(): Promise<ServerMode> {
  const servers = await loadServersRegistry();
  const active = servers.find((s) => s.active);
  if (!active) return 'agent';
  return (active.activeMode === 'admin' ? 'admin' : 'agent');
}

// ---------------------------------------------------------------------------
// Admin cert operations
// ---------------------------------------------------------------------------

/**
 * Check if the active server has an admin certificate configured.
 *
 * Returns true if:
 * - The server has explicit adminAuth, OR
 * - The server has a provider field (cloud-provisioned servers use admin cert as primary)
 */
export async function hasAdminCert(): Promise<boolean> {
  const servers = await loadServersRegistry();
  const active = servers.find((s) => s.active);
  if (!active) return false;

  // Has explicit admin_auth
  if (active.adminAuth) return true;

  // Cloud-provisioned servers (have provider field) use admin cert as primary
  if (active.provider) return true;

  return false;
}

/**
 * Get the admin P12 path for a server, if one exists on disk.
 */
export function getAdminCertPath(serverId: string): string | null {
  const p12Path = serverAdminP12Path(serverId);
  return existsSync(p12Path) ? p12Path : null;
}

/**
 * Get the active server's ID, or null if no active server.
 */
export async function getActiveServerId(): Promise<string | null> {
  const servers = await loadServersRegistry();
  const active = servers.find((s) => s.active);
  return active?.id ?? null;
}

/**
 * Import an admin P12 certificate for a server.
 *
 * - Validates the server ID and source file
 * - Copies P12 to ~/.lamalibre/lamaste/servers/<id>/admin.p12
 * - Updates the server entry in servers.json with adminAuth
 *
 * NOTE: The caller is responsible for storing the P12 password in the OS
 * credential store (this module does not handle credential storage).
 */
export async function importAdminCert(
  serverId: string,
  p12SourcePath: string,
): Promise<{ p12Path: string }> {
  const { mkdir, copyFile, chmod, lstat } = await import('node:fs/promises');
  const pathMod = await import('node:path');

  // Validate server_id is safe for filesystem paths
  if (serverId.includes('/') || serverId.includes('\\') || serverId.includes('\0') || serverId.includes('..')) {
    throw new Error('Server ID contains invalid characters');
  }

  // Validate the source P12 file exists and is a regular file
  const srcStats = await lstat(p12SourcePath);
  if (!srcStats.isFile()) {
    throw new Error('P12 path must be a regular file');
  }

  // Create server directory if needed
  const serverDir = pathMod.join(LAMASTE_DIR, 'servers', serverId);
  await mkdir(serverDir, { recursive: true, mode: 0o700 });

  // Copy P12 to admin.p12
  const dest = pathMod.join(serverDir, 'admin.p12');
  await copyFile(p12SourcePath, dest);
  await chmod(dest, 0o600);

  // Update servers.json with admin_auth
  const servers = await loadServersRegistry();
  const server = servers.find((s) => s.id === serverId);
  if (server) {
    server.adminAuth = {
      method: 'p12',
      p12Path: dest,
    };
    await saveServersRegistry(servers);
  }

  return { p12Path: dest };
}

/**
 * Remove the admin certificate for a server.
 *
 * - Deletes the admin P12 file
 * - Removes adminAuth from servers.json
 * - Resets activeMode to "agent"
 *
 * NOTE: The caller is responsible for cleaning up the credential store entry.
 */
export async function removeAdminCert(serverId: string): Promise<void> {
  const { unlink } = await import('node:fs/promises');

  // Validate server_id is safe for filesystem paths
  if (serverId.includes('/') || serverId.includes('\\') || serverId.includes('\0') || serverId.includes('..')) {
    throw new Error('Server ID contains invalid characters');
  }

  // Delete admin P12 file
  const p12Path = serverAdminP12Path(serverId);
  if (existsSync(p12Path)) {
    await unlink(p12Path);
  }

  // Remove admin_auth from servers.json and reset to agent mode
  const servers = await loadServersRegistry();
  const server = servers.find((s) => s.id === serverId);
  if (server) {
    delete server.adminAuth;
    server.activeMode = 'agent';
    await saveServersRegistry(servers);
  }
}

File: packages/clients/desktop/src-tauri/src/mode.rs
use crate::config;
use crate::credentials;

pub(crate) const AGENTD_PORT: u16 = 9393;

/// Curl the local agentd service (plain HTTP on 127.0.0.1:9393).
///
/// Authenticates with the per-user Bearer token at `~/.lamalibre/lamaste/agentd.token`
/// (mode 0600) — only the agent's owning OS user can read it.
pub(crate) fn curl_agentd(
    method: &str,
    path: &str,
    body: Option<&str>,
) -> Result<String, String> {
    let url = format!("http://127.0.0.1:{}{}", AGENTD_PORT, path);
    let token = crate::api::agentd_token()?;

    let mut args = vec![
        "-s".to_string(),
        "-H".to_string(),
        format!("Authorization: Bearer {}", token),
        "-X".to_string(),
        method.to_string(),
    ];

    if let Some(json_body) = body {
        args.push("-H".to_string());
        args.push("Content-Type: application/json".to_string());
        args.push("-d".to_string());
        args.push(json_body.to_string());
    }

    args.push(url);

    let output = std::process::Command::new("curl")
        .args(&args)
        .output()
        .map_err(|e| format!("Failed to run curl: {}", e))?;

    if !output.status.success() {
        let stderr = String::from_utf8_lossy(&output.stderr);
        return Err(format!("Request failed: {}", stderr));
    }

    let body = String::from_utf8_lossy(&output.stdout).to_string();

    if let Ok(err_obj) = serde_json::from_str::<serde_json::Value>(&body) {
        if let Some(error) = err_obj.get("error").and_then(|e| e.as_str()) {
            return Err(error.to_string());
        }
    }

    Ok(body)
}

/// Set the active mode for a server ("agent" or "admin").
#[tauri::command]
pub async fn set_server_mode(_server_id: String, mode: String) -> Result<(), String> {
    if mode != "agent" && mode != "admin" {
        return Err("Mode must be 'agent' or 'admin'".to_string());
    }

    tokio::task::spawn_blocking(move || {
        let payload = serde_json::json!({ "mode": mode });
        curl_agentd("PATCH", "/api/mode", Some(&payload.to_string()))?;
        Ok(())
    })
    .await
    .map_err(|e| format!("Task failed: {}", e))?
}

/// Get the active server's mode.
#[tauri::command]
pub async fn get_server_mode() -> Result<String, String> {
    tokio::task::spawn_blocking(|| {
        let body = curl_agentd("GET", "/api/mode", None)?;
        let parsed: serde_json::Value = serde_json::from_str(&body)
            .map_err(|e| format!("Failed to parse mode response: {}", e))?;
        parsed
            .get("mode")
            .and_then(|v| v.as_str())
            .map(|s| s.to_string())
            .ok_or_else(|| "Missing 'mode' in response".to_string())
    })
    .await
    .map_err(|e| format!("Task failed: {}", e))?
}

/// Check if the active server has an admin certificate.
#[tauri::command]
pub async fn has_admin_cert() -> Result<bool, String> {
    tokio::task::spawn_blocking(|| {
        let body = curl_agentd("GET", "/api/admin-cert", None)?;
        let parsed: serde_json::Value = serde_json::from_str(&body)
            .map_err(|e| format!("Failed to parse admin-cert response: {}", e))?;
        parsed
            .get("hasAdminCert")
            .and_then(|v| v.as_bool())
            .ok_or_else(|| "Missing 'hasAdminCert' in response".to_string())
    })
    .await
    .map_err(|e| format!("Task failed: {}", e))?
}

/// Import an admin P12 certificate for a server.
#[tauri::command]
pub async fn import_admin_cert(
    server_id: String,
    p12_path: String,
    p12_password: String,
) -> Result<(), String> {
    tokio::task::spawn_blocking(move || {
        // Validate server_id is safe for filesystem paths
        if server_id.contains('/') || server_id.contains('\\') || server_id.contains('\0') || server_id.contains("..") {
            return Err("Server ID contains invalid characters".to_string());
        }

        // Validate the source P12 file exists and is a regular file
        let src = std::path::Path::new(&p12_path);
        let metadata = src.symlink_metadata()
            .map_err(|_| "P12 file not found".to_string())?;
        if !metadata.is_file() {
            return Err("P12 path must be a regular file".to_string());
        }

        // Create server directory if needed
        let server_dir = config::agent_dir().join("servers").join(&server_id);
        std::fs::create_dir_all(&server_dir)
            .map_err(|e| format!("Failed to create server directory: {}", e))?;

        // Copy P12 to admin.p12
        let dest = server_dir.join("admin.p12");
        std::fs::copy(src, &dest)
            .map_err(|e| format!("Failed to copy P12: {}", e))?;

        // Set restrictive permissions
        #[cfg(unix)]
        {
            use std::os::unix::fs::PermissionsExt;
            std::fs::set_permissions(&dest, std::fs::Permissions::from_mode(0o600))
                .map_err(|e| format!("Failed to set permissions: {}", e))?;
        }

        // Store password in credential store (must stay in Rust — OS keychain)
        credentials::store_admin_credential(&server_id, &p12_password)?;

        // Update servers.json via agentd REST API
        let payload = serde_json::json!({
            "adminAuth": {
                "method": "p12",
                "p12Path": dest.to_string_lossy()
            }
        });
        let path = format!("/api/servers/{}", server_id);
        curl_agentd("PATCH", &path, Some(&payload.to_string()))?;

        Ok(())
    })
    .await
    .map_err(|e| format!("Task failed: {}", e))?
}

/// Remove the admin certificate for a server.
#[tauri::command]
pub async fn remove_admin_cert(server_id: String) -> Result<(), String> {
    tokio::task::spawn_blocking(move || {
        // Validate server_id is safe for filesystem paths
        if server_id.contains('/') || server_id.contains('\\') || server_id.contains('\0') || server_id.contains("..") {
            return Err("Server ID contains invalid characters".to_string());
        }

        // Delete admin P12 file
        let admin_p12 = config::agent_dir().join("servers").join(&server_id).join("admin.p12");
        if admin_p12.exists() {
            std::fs::remove_file(&admin_p12)
                .map_err(|e| format!("Failed to remove admin P12: {}", e))?;
        }

        // Delete credential (must stay in Rust — OS keychain)
        let _ = credentials::delete_admin_credential(&server_id);

        // Remove adminAuth and reset mode via agentd REST API
        // Use null to remove the adminAuth field, and reset mode to agent
        let payload = serde_json::json!({
            "adminAuth": null,
            "activeMode": "agent"
        });
        let path = format!("/api/servers/{}", server_id);
        curl_agentd("PATCH", &path, Some(&payload.to_string()))?;

        Ok(())
    })
    .await
    .map_err(|e| format!("Task failed: {}", e))?
}
